README.md
<div align="center">
  <img src="https://discordmods.cmtapp.fr/discordmods.jpg" height="300px"><br>
  <a href="https://www.npmjs.com/package/discord.mods"><img src="https://img.shields.io/npm/v/discord.mods?style=for-the-badge" alt="Version" /></a>  
  <a href="https://www.npmjs.com/package/discord.mods"><img src="https://img.shields.io/npm/dm/discord.mods?style=for-the-badge" alt="Downloads" /></a><br>
  <a href="https://www.npmjs.com/package/discord.mods"><img src="https://nodei.co/npm/discord.mods.png?downloads=true&stars=true" alt="npm installInfo" /></a>
</div>

# Contents

- [Installation](#installation)
- [Usage](#usage)
  - [Authentification](#authentification)
  - [Action](#action)
  - [Tools](#tools)
  - [Arter](#arter)
- [Links](#links)
- [Help](#help)
- [Register Your Bot](#register-bot)

# Installation

Execute `npm i discord.mods`, and then install "BetterDiscord"...<br> And betterdiscord plugin for discord.mods ... [Links](#links)<br>
It's Ready!

# Usage
How to use ? It's THE question.<br>

## Authentification

```js
const Discord = require('discord.js')
const client = new Discord.Client()

var DM = require('./packages/index')
var DiscordMods = new DM();

client.login("Super-Secret-Token").then(()=>{
    DiscordMods.login(client, "Super-Secret-DiscordMods-Token")
})
```

```js
DiscordMods.action // Return Action Class
DiscordMods.tools // Return Tools Class
DiscordMods.arter // Return Arter Class
```

## Action

### Alert
Create an alert allows you to display a pop-up in an efficient and simple way

```js
DiscordMods.action.alert(Discord.Channel `or` Discord.Guild, Discord.User, {
  title: "Your Super title",
  subtitle: "The super content :D"
})
```

<img src="https://discordmods.cmtapp.fr/example1.png" height="150px">

### Toast
Create a notification

```js
DiscordMods.action.toast(Discord.Channel `or` Discord.Guild, Discord.User, {
  title: "A ULTRA SUPER MEGA GOD TITLE OMG... and yes... i'm fine",
  type: "danger", // "danger", "success", "warn", "info"
  icon: false, // Active icon in the toast ?
  timeout: 5000 // How many time alive ?
})
```

### html
Create an empty pop-up, which can contain html elements

```js
DiscordMods.action.html(Discord.Channel `or` Discord.Guild, Discord.User, {
  html: `html? OK... <img src="i dont know">`
})
```

And if you will change size of window... Color Background of close button... Text of close button...<br>
MaxSize is 2 thirds of the discord window pixel value (separately for width and height)<br>
MinSize is 0px<br>
A fix size is not allowed

```js
DiscordMods.action.html(Discord.Channel `or` Discord.Guild, Discord.User, {
  html: `<img src="https://discordmods.cmtapp.fr/example2.png" height="150px">`,
  minWidth: 400, // you can change "min" to "max"
  minHeight: 200,
  button: "Don't Touch Me",
  color: "magenta"
})
```

<img src="https://discordmods.cmtapp.fr/example3.png" height="150px">

ExampleCeption D:

## Tools

### isConnected (userId)
Check if user is Connected with DiscordMods

```js
await DiscordMods.tools.isConnected(Discord.User.id) // Return true if the user is connected else return false
```

### escapeHtml (html)
Removes html execution from a text containing html

```js
DiscordMods.tools.escapeHtml("<img src='a super image'>") // Instead of displaying the image it displays : "<img src='a super image'>"
```

### markdown (string)
Convert text to markdown

```js
DiscordMods.tools.markdown("**test**") // Return the markdowned text!
```

## Arter



# Links

- <a href="https://discordmods.cmtapp.fr/api?v=2&r=download">Download Plugins</a><br>
- <a href="https://discordmods.cmtapp.fr/">Web</a><br>
- <a href="https://github.com/YiraSan/discord.mods">GitHub</a><br>
- <a href="https://discord.gg/4QwrJmj">Discord</a>

# Help

For Help go to my Discord :D ... [Links](#links)<br> 
And issues : https://github.com/YiraSan/discord.mods/issues

# Register Bot

Reback soon!

CHANGELOG.md
# 0.4

- All Remaded.
- All Improved.
- Now using socket.io 3.1
- Auto-Reconnecting
- inputHtml now html.
- isConnected require now an async await.
```js
await DiscordMods.tools.isConnected(Discord.User.id) // Return true or false
```
- Arter added.
- Support french & english
- Securised Interface (ssl connection)

# 0.3.30 [Not Released, Edit Saved]

- Add Session System
- Catch/Then/Finnally hotfix

# 0.3.25

- For includes markdown in inputHtml use the tools function : "markdown"
```js
DiscordMods.tools.markdown("**test**") // Return the markdowned text!
```

# 0.3.2

- title in inputHtml are delete
- Change color of the "inputHtml" button (with "color" parameters)
- Change text of the "inputHtml" button (with "text" parameters)
- Change size of the "inputHtml" window (show example, size in pixel)
- HotFix on action displayer

```js
DiscordMods.action.inputHtml({
    html: "HTML HERE!",
    color: "yellow", // WoW! A yellow button! (#?????? it's valid) 
    text: "Don't Touch Me",
    maxWidth: 800, // Max is 2 thirds of the discord window pixel value (separately for width and height)
    maxHeight: 800,
    minWidth: 100,
    minWidth: 100
})
```

# 0.3.1

- API v2 it's here! (upgrade speed, fluidity & multi-support)
- The Action "inputHtml" add! (create a Window with your html) [WARNING: IF YOU INCLUDES `<script>` in your html, the client plugins execute escapeHtml function on your content/html]
- Add "escapeHtml" in Tools : The function is used to prevent the execution of html in content running html. Useful if you include user content (for security)

```js
DiscordMods.action.inputHtml({
    html: "HTML HERE!",
    title: "A title?" // title parameters is optionnal (add title in window, as Alert), title don't support html!
})
```

- Patch Error Log

# 0.3.0

- Catch/Then & Finnally is here
- Restructure Packages
- Preparing for API v2 compatibility
- Add Custom POST Json Request in Tools

```js
DiscordMods.tools.request({
    user: "you or me ? (id)",
    channel: "The channel or guild?",
    type: 'alert', // "toast"/"alert"/"html"
    params: {
        title: "Your Super Title!",
        subtitle: "The content of alert?"
    }  
})
```

- The Packages is only in english now

# 0.2.9

- Initial Commit in npm.js & github

# 0.2.8 (and lower)

Which gave the 0.2.8

lib/errorReader.js
module.exports = function errorReader(error, api) {
    if (error == 0) {
        console.log(`[DiscordMods] Cannot connect with the API (ApiResponse: ${api}) [Code: ${error}]`)
    } else if (error == 4000) {
        console.log(`[DiscordMods] Information is missing from your request (ApiResponse: ${api}) [Code: ${error}]`)
    } else if (error == 4001) {
        console.log(`[DiscordMods] User is not logged in (ApiResponse: ${api}) [Code: ${error}]`)
    } else if (error == 3000) {
        console.log(`[DiscordMods] The API refused your request (ApiResponse: ${api}) [Code: ${error}]`)
    } else if (error == 3002) {
        console.log(`[DiscordMods] The API to return that the query did not exist (ApiResponse: ${api}) [Code: ${error}]`)
    } else if (error == 200) {
        // ITS OK
    } else if (error == 201) {
        // ITS OK
    } else {
        console.log(`[DiscordMods] An unknown error has occurred (ApiResponse: ${api}) [Code: ${error}]`)
    }

    return;
}

index.js
// LIB IMPORT

var Return = require('./lib/Return')

function pfs(ident="**", dif="strong", c="") {

    var c2 = c.split(ident)

    if (c2.length < 3) return c;

    var final = "";

    for (let i = 1; i < c2.length; i+=2) {

        final += c2[i-1]
        final += `<${dif}>${c2[i]}</${dif}>`
        final += c2[i+1]

    }

    return final;

}

class Action {

    constructor (socket) {
        this.socket = socket;
    }

    toast(to, user, object={
        title: "Work!",
        icon: true,
        type: "success",
        timeout: 4000,
    }) {
        
        var data = {
            type: "push",
            to: to.id,
            object: object,
        }

        this.socket.emit("request", "json", user.id, data);
        
        return new Return(true);

    }

    alert(to, user, object={
        title: "Example",
        subtitle: "Work too!"
    }) {
        
        var data = {
            type: "alert",
            to: to.id,
            object: object,
        }

        this.socket.emit("request", "json", user.id, data);
        
        return new Return(true);

    }

    html(to, user, object={
        html: `Touch my Tralala`,
        minWidth: 400, 
        minHeight: 200,
        maxWidth: 400, 
        maxHeight: 200,
        button: "My Ding Ding Dong",
        color: "magenta"
    }) {
        
        var data = {
            type: "html",
            to: to.id,
            object: object,
        }

        this.socket.emit("request", "json", user.id, data);
        
        return new Return(true);

    }

}

class Tools {

    constructor (socket) {
        this.socket = socket;
    }

    escapeHtml(text) {
        var map = {
          '&': '&amp;',
          '<': '&lt;',
          '>': '&gt;',
          '"': '&quot;',
          "'": '&#039;'
        };
        
        return text.replace(/[&<>"']/g, function(m) { return map[m]; });
    }

    markdown(content) {
        var content = pfs("**", "strong", content)
        var content = pfs("*", "em", content)
        var content = pfs("__", "u", content)
        var content = pfs("_", "em", content)
        var content = pfs("~~", "s", content)
        return content;
    }

    isConnected(id) {

        return new Promise(resolve => {
          
            var f = (...args) => {
                if (args[0] == "online") {
                    resolve(args[1])
                    this.socket.off("response", f)
                }
            }

            this.socket.on("response", f)

            this.socket.emit("request", "online", id)
            
        })
        
    }

}

class Arter {

    /** 
    * @name Arter
    * @author YiraSan
    * @version 1.0.5
    * @description Complete personnalisation of your themes
    */

    constructor(socket) {
        this.socket = socket;
    }

    editCSS(to, user, object={
        popoutModalBrightness: 0.75,
        popoutModalImage: `var(--background-image)`,
        popoutModalBlur: `5px`,
        popoutModalSize: `cover`,
        popoutModalPosition: `center`,
        backgroundColor: `transparent`,
        backgroundImage: `url('https://arter.cmtapp.fr/?img')`,
        backgroundImageBlur: `5px`,
        windowPadding: `0px`,
        windowRoundess: `0px`,
        textShadow: 1,
        embedBorder: `none`,
        colorPrimary: `252,68,18`,
        colorSecondary: `252,132,18`,
        colorDirection: `320deg`,
        scrollbarColor: `rgba(255,255,255,0.05)`,
        linkColor: `#00b0f4`,
    
    }) {
        
        var data = {
            type: "arter",
            to: to.id,
            object: object,
        }

        this.socket.emit("request", "json", user.id, data);
        
        return new Return(true);

    }

}

module.exports = class {

    constructor() {
        this.socket = require('socket.io-client')("https://dms.cmtchat.online/");

        this.socket.on("disconnect", () => {
            console.log("[DiscordMods] Disconnected of DiscordMods Server")
        })
    }

    get tools() {
        return new Tools(this.socket);
    }

    get action() {
        return new Action(this.socket);
    }

    get arter() {
        return new Arter(this.socket);
    }

    login(client, token) {

        var id = client.user.id;

        this.socket.emit("user", "bot", id)
        this.socket.emit("token", token)

        this.socket.on("connect", ()=>{
            this.socket.emit("user", "bot", id)
            this.socket.emit("token", token)
            console.log("[DiscordMods] Reconnected to the DiscordMods Server")
        })

        return new Return(false, this);
    }

}